iroh-net/src/util/watchable.rs

```diff
@@ -0,0 +1,401 @@
+//! Watchable values.
+//!
+//! A [`Watchable`] exists to keep track of a value which may change over time.  It allows
+//! observers to be notified of changes to the value.  The aim is to always be aware of the
+//! **last** value, not to observe every value there has ever been.
+
+use std::collections::VecDeque;
+use std::future::{self, Future};
+use std::pin::Pin;
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::sync::{Arc, RwLock};
+use std::task::{self, Poll, Waker};
+
+use futures_lite::stream::Stream;
+
+const INITIAL_EPOCH: u64 = 1;
+
+/// The shared state for a [`Watchable`].
+#[derive(Debug)]
+struct Shared<T> {
+    /// The value to be watched.
+    ///
+    /// Note that the `Option` is only there to allow initialisation.  Once initialised the
+    /// value can never be cleared again.
+    value: RwLock<Option<T>>,
+    epoch: AtomicU64,
+    watchers: RwLock<VecDeque<Waker>>,
+}
+
+impl<T> Default for Shared<T> {
+    fn default() -> Self {
+        Shared {
+            value: Default::default(),
+            epoch: INITIAL_EPOCH.into(),
+            watchers: Default::default(),
+        }
+    }
+}
+
+impl<T: Clone> Shared<T> {
+    /// Returns the value, initialised or not.
+    fn get(&self) -> Option<T> {
+        self.value.read().unwrap().clone()
+    }
+
+    /// Returns a future completing once the value is initialized.
+    ///
+    /// If the value is already initialised the future will complete immediately.
+    // TODO: maybe writing this as a poll function avoids needing to use Either?
+    fn initialized(&self) -> impl Future<Output = T> + '_ {
+        let epoch = self.epoch.load(Ordering::Acquire);
+        if let Some(ref value) = *self.value.read().expect("poisoned") {
+            return Either::Left(future::ready(value.clone()));
+        }
+        Either::Right(future::poll_fn(move |cx| {
+            self.poll_next(cx, epoch).map(|(_, t)| t)
+        }))
+    }
+
+    fn poll_next(&self, cx: &mut task::Context<'_>, last_epoch: u64) -> Poll<(u64, T)> {
+        let epoch = self.epoch.load(Ordering::Acquire);
+        // TODO(flub): Pretty sure this can miss a write because the epoch and wakers are
+        //    separate:
+        //    - thread 1 runs poll_next
+        //    - thread 2 runs set
+        //    1. thread 1: load epoch
+        //    2. thread 2: lock value, replace value, unlock value
+        //    3. thread 2: store epoch
+        //    4. thread 2: lock wakers, drain wakers, unlock wakers
+        //    5. thread 1: lock wakers, install waker, unlock wakers
+        //
+        //    I believe the epoch and wakers need to be stored in the same RwLock.
```

Comment:
You're completely right about the wakers possibly missing a wakeup once a value is available.
I've tested this implementation of Watchable with [`loom`](https://docs.rs/loom/latest/loom/), and can confirm that it gets stuck waiting for `watcher.initialized()` in some cases, even though there's a `watchable.set()` call.
(Interestingly, reproducing it practically *requires* loom usage, since it's very timing-sensitive and *doesn't* reproduce easily with running the same test case in a tight loop a couple thousand times.)

You're incorrect about the fact that the epoch and wakers needing to be stored on the same RwLock.

Similar to [this example](https://docs.rs/futures/latest/futures/task/struct.AtomicWaker.html#examples), after writing the waker, you can re-check the epoch to see if there was a write in-between.

Adding this second check, it passes loom.